### PR TITLE
418 query pre logic architecture

### PR DESCRIPTION
**Pull request by ElliotAOram**
Mid way review

### CODE REVIEW

File: scripts/system_performance/run_statistics.py [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+        {instrument_name : [[method output row 1]],[[method output row 2] etc] ...}"""
+        for sublist in self.list_all_instruments():
+            try:
+                if method_name in self.create_method_mappings():
+                    instrument_dict[sublist[1]] = self.create_method_mappings()[method_name](sublist[0])
+            except KeyError:
+                raise ValueError('Invalid Input - method {} does not exist'.format(method_name))
+        return instrument_dict
+
+    def input_verification(self, instrument_input, method_name):
```
Comment (ElliotAOram) [inline scripts/system_performance/run_statistics.py]: I would suggest that `input_verification` is a somewhat misleading function name as the function does more than just verification.

File: scripts/system_performance/run_statistics.py [hunk 2]
```diff
@@ -0,0 +1,292 @@
...
+        """Checks that the instrument's in method input exist and then calls
+        and returns the methods used as input for each instrument placing in
+        a dictionary as a nested list for each instrument as:
+         {instrument_name : [[method output row 1]],[[method output row 2] etc.] ...}"""
```
Comment (ElliotAOram) [inline scripts/system_performance/run_statistics.py]: This needs some rewording as it's very difficult to follow I would suggest adding doxygen style comments to state the params and their types and simplifying the function description:
```
Run methods for each instrument name
:param instrument_input: [(str), (str), ...] - Represents the instruments to be used in queries
:param method_name: (str) -  The name of the methods to run
:return: A dictionary of instrument name (key) and list of method output (value)
```

File: scripts/system_performance/run_statistics.py [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+        instrument_dict = {}
+
+        for sublist in self.list_all_instruments():
```
Comment (ElliotAOram) [inline scripts/system_performance/run_statistics.py]: what is sublist here?

File: scripts/system_performance/run_statistics.py [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+                in_cycle = None
+                return in_cycle
+        # Looking at last row in data frame. Current date can't be checked any further back in time
+        else:
+            in_cycle = None
+            return in_cycle
+
+    def create_method_mappings(self):
+        """A dictionary to map on input, user specified methods to return to equivalent method to be called"""
+        method_mappings = {'missing_run_numbers_report': self.missing_run_numbers_report,
```
Comment (ElliotAOram) [inline scripts/system_performance/run_statistics.py]: nit: can just return dict and save a line
```
return {'missing...': self...,
        ...}
```

File: scripts/system_performance/run_statistics.py [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+         {instrument_name : [[method output row 1]],[[method output row 2] etc.] ...}"""
+        instrument_dict = {}
+
+        for sublist in self.list_all_instruments():
+            # if input is None, run all instruments by default
+            if not isinstance(instrument_input, list):
+                logging.error("Value is not iterable, the first argument must be of type list")
+                return None
+            for instrument in instrument_input:
+                if sublist[1] == instrument:
```
Comment (ElliotAOram) [inline scripts/system_performance/run_statistics.py]: only ever using `sublist[1]`. Is this correct functionality

File: scripts/system_performance/run_statistics.py [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+                    try:
+                        if method_name in self.create_method_mappings():
+                            instrument_dict[sublist[1]] = self.create_method_mappings()[method_name](sublist[0])
+                    except KeyError:
+                        raise ValueError('Invalid Input - method {} does not exist'.format(method_name))
+                # Run all methods is user specifies "all"
+                elif instrument == 'all' or not instrument:
+                    self.run_every_instruments(instrument_dict, method_name)
+                else:
+                    logging.info('The instrument: {} is not currently using Autoreduction'.format(instrument))
```
Comment (ElliotAOram) [inline scripts/system_performance/run_statistics.py]: `The instrument: {} has not been found in the autoreduction database`
The above is more obvious what the error is for developers (which is the main audience for this error)

File: scripts/system_performance/run_statistics.py [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+    @staticmethod
+    def list_all_instruments():
+        """Returns a list of all currently active instruments"""
+        instrument_list = DatabaseMonitorChecks.list_instruments()
+        return instrument_list
+
+    def run_every_instruments(self, instrument_dict, method_name):
+        """Returns all existing instruments in a dictionary for a given method as:
+        {instrument_name : [[method output row 1]],[[method output row 2] etc] ...}"""
+        for sublist in self.list_all_instruments():
```
Comment (ElliotAOram) [inline scripts/system_performance/run_statistics.py]: better name for `sublist`

File: scripts/system_performance/run_statistics.py [hunk 8]
```diff
@@ -0,0 +1,292 @@
...
+        method_mappings = {'missing_run_numbers_report': self.missing_run_numbers_report,
+                           'execution_time': self.execution_time,
+                           'execution_time_average': self.execution_time_average,
+                           'run_frequency': self.run_frequency,
+                           'run_frequency_average': self.run_frequency_average
+                           }
+        return method_mappings
+
+    @staticmethod
+    def list_all_instruments():
```
Comment (ElliotAOram) [inline scripts/system_performance/run_statistics.py]: rename to `get_instrument_models` or similar

**Comment by JackEAllen** [comment]
All refactors have been made 👍

**Pull request closed without merging**